File: backend/app/schemas/business.py
from datetime import datetime
from typing import Any, Literal

from pydantic import (
    AliasChoices,
    BaseModel,
    ConfigDict,
    EmailStr,
    Field,
    field_validator,
    model_validator,
)

from app.models.enums import (
    BusinessClaimStatus,
    BusinessMembershipRole,
    BusinessVerificationStatus,
)


PublicBusinessStatus = Literal["unverified", "claimed", "verified"]


class BusinessCreate(BaseModel):
    slug: str | None = Field(None, min_length=2, max_length=160, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    name: str = Field(..., min_length=2, max_length=200)
    description: str | None = Field(None, max_length=5000)
    compound_id: int | None = Field(None, ge=1)
    city: str = Field(..., min_length=2, max_length=120)
    area: str | None = Field(None, max_length=120)
    category: str = Field(..., min_length=2, max_length=120)
    address: str | None = Field(None, max_length=1000)
    phone: str | None = Field(None, min_length=7, max_length=32)
    whatsapp: str | None = Field(None, min_length=7, max_length=32)
    email: EmailStr | None = None
    website: str | None = Field(None, max_length=500)
    contact_name: str | None = Field(None, max_length=200)
    hours: dict[str, Any] = Field(default_factory=dict)
    verification_status: BusinessVerificationStatus = BusinessVerificationStatus.UNVERIFIED
    is_active: bool = True
    is_hidden: bool = False

    @field_validator("name", "city", "category", "area", "address", "contact_name", mode="before")
    @classmethod
    def strip_text(cls, value: str | None) -> str | None:
        return value.strip() if isinstance(value, str) else value

    @field_validator("website")
    @classmethod
    def validate_website(cls, value: str | None) -> str | None:
        if value and not value.lower().startswith(("http://", "https://")):
            raise ValueError("website must use http:// or https://")
        return value


class BusinessUpdate(BaseModel):
    slug: str | None = Field(None, min_length=2, max_length=160, pattern=r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
    name: str | None = Field(None, min_length=2, max_length=200)
    description: str | None = Field(None, max_length=5000)
    compound_id: int | None = Field(None, ge=1)
    city: str | None = Field(None, min_length=2, max_length=120)
    area: str | None = Field(None, max_length=120)
    category: str | None = Field(None, min_length=2, max_length=120)
    address: str | None = Field(None, max_length=1000)
    phone: str | None = Field(None, min_length=7, max_length=32)
    whatsapp: str | None = Field(None, min_length=7, max_length=32)
    email: EmailStr | None = None
    website: str | None = Field(None, max_length=500)
    contact_name: str | None = Field(None, max_length=200)
    hours: dict[str, Any] | None = None
    verification_status: BusinessVerificationStatus | None = None
    is_active: bool | None = None
    is_hidden: bool | None = None

    @model_validator(mode="after")
    def require_update(self):
        if not self.model_fields_set:
            raise ValueError("at least one field must be supplied")
        return self

    @field_validator("website")
    @classmethod
    def validate_website(cls, value: str | None) -> str | None:
        if value and not value.lower().startswith(("http://", "https://")):
            raise ValueError("website must use http:// or https://")
        return value


class BusinessResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    slug: str
    name: str
    description: str | None
    compound_id: int | None
    city: str
    area: str | None
    category: str
    address: str | None
    phone: str | None
    whatsapp: str | None
    email: str | None
    website: str | None
    contact_name: str | None
    hours: dict[str, Any]
    verification_status: BusinessVerificationStatus
    public_status: PublicBusinessStatus
    is_active: bool
    is_hidden: bool
    viewer_claim_status: BusinessClaimStatus | None = None
    viewer_membership_role: BusinessMembershipRole | None = None
    created_at: datetime
    updated_at: datetime


class BusinessListResponse(BaseModel):
    items: list[BusinessResponse]
    total: int
    skip: int
    limit: int


class BusinessClaimCreate(BaseModel):
    full_name: str = Field(..., min_length=2, max_length=200)
    relationship_role: str = Field(..., min_length=2, max_length=120)
    phone: str = Field(..., min_length=7, max_length=32)
    email: EmailStr
    supporting_info: str | None = Field(
        None,
        max_length=5000,
        validation_alias=AliasChoices("supporting_info", "supporting_information"),
    )
    supporting_documents: list[str] = Field(default_factory=list, max_length=20)

    @field_validator("full_name", "relationship_role", "phone", mode="before")
    @classmethod
    def strip_required(cls, value: str) -> str:
        return value.strip()

    @field_validator("supporting_documents")
    @classmethod
    def validate_documents(cls, values: list[str]) -> list[str]:
        if any(not value.strip() or len(value) > 1000 for value in values):
            raise ValueError("supporting document references must be non-empty and at most 1000 characters")
        return values


class BusinessClaimReview(BaseModel):
    review_notes: str | None = Field(None, max_length=5000)
    membership_role: BusinessMembershipRole = BusinessMembershipRole.OWNER


class BusinessClaimResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    business_id: int
    business_slug: str
    business_name: str
    business_verification_status: BusinessVerificationStatus
    public_status: PublicBusinessStatus
    claimant_id: int | None
    full_name: str
    relationship_role: str
    phone: str
    email: str
    supporting_info: str | None
    supporting_documents: list[str]
    status: BusinessClaimStatus
    submitted_at: datetime
    reviewed_at: datetime | None
    reviewer_id: int | None
    review_notes: str | None


class BusinessClaimListResponse(BaseModel):
    items: list[BusinessClaimResponse]
    total: int
    skip: int
    limit: int


class BusinessSearchResult(BaseModel):
    id: int
    slug: str
    name: str
    city: str
    area: str | None
    category: str
    verification_status: BusinessVerificationStatus
    public_status: PublicBusinessStatus

File: packages/shared/src/schemas/business.ts
import { z } from "zod";

export const BusinessVerificationStatusSchema = z.enum([
  "UNVERIFIED",
  "CLAIMED",
  "VERIFIED",
]);
export const BusinessClaimStatusSchema = z.enum(["PENDING", "APPROVED", "REJECTED"]);
export const BusinessMembershipRoleSchema = z.enum(["OWNER", "MANAGER"]);

export const BusinessHoursDaySchema = z.object({
  open: z.string().optional(),
  close: z.string().optional(),
  closed: z.boolean().optional(),
});

export const BusinessHoursSchema = z.record(z.string(), BusinessHoursDaySchema);

export const BusinessOfferCreateSchema = z.object({
  title: z.string().trim().min(1).max(200),
  description: z.string().trim().max(2000).optional(),
  badge_text: z.string().trim().max(80).optional(),
  starts_at: z.string().datetime().optional(),
  ends_at: z.string().datetime().optional(),
  is_active: z.boolean().optional(),
});

export const BusinessOfferUpdateSchema = BusinessOfferCreateSchema.partial();

export const BusinessOfferSchema = BusinessOfferCreateSchema.extend({
  id: z.number().int().positive(),
  business_id: z.number().int().positive(),
  description: z.string().nullable().optional(),
  badge_text: z.string().nullable().optional(),
  starts_at: z.string().datetime().nullable().optional(),
  ends_at: z.string().datetime().nullable().optional(),
  is_active: z.boolean(),
  created_at: z.string().datetime().optional(),
  updated_at: z.string().datetime().optional(),
});

export const BusinessAnalyticsSchema = z.object({
  profile_views: z.number().int().nonnegative(),
  offer_clicks: z.number().int().nonnegative(),
  active_offers: z.number().int().nonnegative(),
});

export const BusinessSummarySchema = z.object({
  id: z.number().int().positive(),
  slug: z.string(),
  name: z.string(),
  category: z.string(),
  city: z.string().nullable().optional(),
  area: z.string().nullable().optional(),
  compound_id: z.number().int().positive().nullable().optional(),
  compound_name: z.string().nullable().optional(),
  image_url: z.string().url().nullable().optional(),
  verification_status: BusinessVerificationStatusSchema,
  is_active: z.boolean(),
});

export const BusinessDetailSchema = BusinessSummarySchema.extend({
  description: z.string().nullable().optional(),
  address: z.string().nullable().optional(),
  phone: z.string().nullable().optional(),
  email: z.string().email().nullable().optional(),
  website: z.string().url().nullable().optional(),
  hours: BusinessHoursSchema.nullable().optional(),
  is_hidden: z.boolean().optional(),
  user_membership_role: BusinessMembershipRoleSchema.nullable().optional(),
  viewer_membership_role: BusinessMembershipRoleSchema.nullable().optional(),
  current_user_claim_status: BusinessClaimStatusSchema.nullable().optional(),
  profile_views: z.number().int().nonnegative().optional(),
  offers: z.array(BusinessOfferSchema).default([]),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
});

export const BusinessCreateSchema = z.object({
  name: z.string().trim().min(2).max(200),
  category: z.string().trim().min(1).max(100),
  description: z.string().trim().max(4000).optional(),
  compound_id: z.number().int().positive().optional(),
  city: z.string().trim().max(100).optional(),
  area: z.string().trim().max(100).optional(),
  address: z.string().trim().max(500).optional(),
  phone: z.string().trim().max(50).optional(),
  email: z.string().email().optional(),
  website: z.string().url().optional(),
  image_url: z.string().url().optional(),
  hours: BusinessHoursSchema.optional(),
});

export const BusinessDirectoryResponseSchema = z.object({
  items: z.array(BusinessSummarySchema),
  total: z.number().int().nonnegative(),
  skip: z.number().int().nonnegative(),
  limit: z.number().int().positive(),
});

export const BusinessClaimCreateSchema = z.object({
  full_name: z.string().trim().min(2).max(200),
  relationship_role: z.string().trim().min(2).max(100),
  phone: z.string().trim().min(5).max(50),
  email: z.string().email(),
  supporting_information: z.string().trim().max(4000).optional(),
  supporting_info: z.string().trim().max(5000).optional(),
  supporting_documents: z.array(z.string()).max(20).optional(),
  requested_role: BusinessMembershipRoleSchema.default("OWNER"),
});

export const BusinessClaimSchema = BusinessClaimCreateSchema.extend({
  id: z.number().int().positive(),
  business_id: z.number().int().positive(),
  business_slug: z.string().optional(),
  business_name: z.string().optional(),
  business_verification_status: BusinessVerificationStatusSchema.optional(),
  public_status: z.string().optional(),
  claimant_user_id: z.number().int().positive().optional(),
  claimant_id: z.number().int().positive().nullable().optional(),
  status: BusinessClaimStatusSchema,
  submitted_at: z.string().datetime(),
  reviewed_at: z.string().datetime().nullable().optional(),
  reviewed_by_id: z.number().int().positive().nullable().optional(),
  reviewer_id: z.number().int().positive().nullable().optional(),
  review_notes: z.string().nullable().optional(),
});

export const BusinessClaimReviewSchema = z.object({
  status: z.enum(["APPROVED", "REJECTED"]),
  review_notes: z.string().trim().max(2000).optional(),
  membership_role: BusinessMembershipRoleSchema.optional(),
});

export const BusinessMembershipSchema = z.object({
  id: z.number().int().positive(),
  business_id: z.number().int().positive(),
  user_id: z.number().int().positive(),
  role: BusinessMembershipRoleSchema,
  created_at: z.string().datetime(),
});

export const AdminBusinessUpdateSchema = BusinessCreateSchema.partial().extend({
  verification_status: BusinessVerificationStatusSchema.optional(),
  is_active: z.boolean().optional(),
  is_hidden: z.boolean().optional(),
});

export type BusinessHoursDay = z.infer<typeof BusinessHoursDaySchema>;
export type BusinessHours = z.infer<typeof BusinessHoursSchema>;
export type BusinessVerificationStatus = z.infer<typeof BusinessVerificationStatusSchema>;
export type BusinessClaimStatus = z.infer<typeof BusinessClaimStatusSchema>;
export type BusinessMembershipRole = z.infer<typeof BusinessMembershipRoleSchema>;
export type BusinessSummary = z.infer<typeof BusinessSummarySchema>;
export type BusinessDetail = z.infer<typeof BusinessDetailSchema>;
export type BusinessCreate = z.infer<typeof BusinessCreateSchema>;
export type BusinessDirectoryResponse = z.infer<typeof BusinessDirectoryResponseSchema>;
export type BusinessClaimCreate = z.infer<typeof BusinessClaimCreateSchema>;
export type BusinessClaim = z.infer<typeof BusinessClaimSchema>;
export type BusinessClaimReview = z.infer<typeof BusinessClaimReviewSchema>;
export type BusinessMembership = z.infer<typeof BusinessMembershipSchema>;
export type AdminBusinessUpdate = z.infer<typeof AdminBusinessUpdateSchema>;
export type BusinessOffer = z.infer<typeof BusinessOfferSchema>;
export type BusinessOfferCreate = z.infer<typeof BusinessOfferCreateSchema>;
export type BusinessOfferUpdate = z.infer<typeof BusinessOfferUpdateSchema>;
export type BusinessAnalytics = z.infer<typeof BusinessAnalyticsSchema>;
